Documentation: Basic Operators

iv-mexx opened the pull request:
This adds descriptions and examples for the basic operators to the existing stub. 

I have tried to keep the examples short and similar to the [legacy ObjC Version of this Document](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/swift-development/Documentation/Legacy/BasicOperators.md#injecting-effects)

The examples have been written and tested in a playground, so they should work correctly.

I was also playing around with the idea of providing these examples in a playground, but how that would work specifically is not clear to me (Needs extra workspace with RAC Framework built, how to separate the different Examples?). If you like the idea, I can play around with that later in a separate PR. then again, the rest of the still missing documentation may be more important...

This refs #1486

neilpa commented:
> I was also playing around with the idea of providing these examples in a playground

Long term this would be cool, although I wouldn't worry too much about it now. There are also [tools](https://github.com/jas/playground) for generating playgrounds from markdown docs.

JaviSoto commented:
Xcode 6.3 supports adding Markdown to playgrounds directly :) I think that would be a really good approach. A "getting started" playground for ReactiveCocoa showcasing the most common things you do with it would serve as a really great way to understand the basics while playing with it interactively :+1:

NachoSoto commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,362 @@ code-style will be used.
+
 ### Combining latest values
+
+The `combineLatest` function combines the latest values of two (or more) signals. The resulting signal will not send a value until both inputs have sent at least one value each.
```
I would mention that that is only true for the first value, but new values are always sent when the input signals emit a new value (vs `zip`).

iv-mexx replied:
Ok, I've tried to change the second sentence to
> The resulting signal will only send a the first value after both inputs have sent at least one value each. After that, each value on either of the inputs will cause a new value on the output.

And also added a counterpart to `zip`:

> The elements of any Nth pair are the Nth elements of the two input signals. That means the output signal will always wait for both input signals to send and output.

NachoSoto commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,362 @@ code-style will be used.
+sendNext(lettersSink, "A")  // prints (1, A)
+sendNext(numbersSink, 2)    // prints (2, A)
+sendNext(lettersSink, "B")  // prints (2, B)
+sendNext(lettersSink, "C")  // prints (2, C)
```
Maybe sendCompleted as well to illustrate when the resulting `signal` completes?

iv-mexx replied:
Good point

NachoSoto commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,362 @@ code-style will be used.
+sendNext(lettersSink, "A")  // prints (0, A)
+sendNext(numbersSink, 2)    // nothing printed
+sendNext(lettersSink, "B")  // prints (1, B)
+sendNext(lettersSink, "C")  // prints (2, C)
```
Ditto: maybe `sendCompleted` as well to illustrate when the resulting `signal` completes?

NachoSoto commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,362 @@ code-style will be used.
 
-### Concatenating
+The `flatten` operators transforms a `SignalProducer`-of-`SignalProducer`s into a single `SignalProducer`.
+There are multiple different semantics of the operation which can be chosen as a `FlattenStrategy`.
```
I wonder if it would be helpful to illustrate first why flattening can be done in several ways, by showing an example with an array of arrays.

iv-mexx replied:
Are you thinking about just a plain array of arrays, no signals involved?

```
let array = [[1], [2, 3], [4, 5, 6]]
let flattened = arr.flatMap{ $0 }
println(flattened) // prints [1, 2, 3, 4, 5, 6]
```

I'm not sure how to demonstrate the different strategies that way, I guess I misunderstand what you mean :see_no_evil:

neilpa replied:
I think he means a way to visualize how values interleave with the different strategies.

```swift
let values = [
// imagine column offset as time
[ 1,    2,      3 ],
   [ 4,      5,     6 ],
         [ 7,     8,   9 ],
]

let concat = 
[ 1,    2,      3,4,      5,     6,7,     8,   9]

let latest =
[ 1, 4,    7,     8,   9 ]

let merge =
[ 1, 4, 2, 7,5, 3,8,6, 9 ]
```

This may be a bit much but I've always found [operator diagrams](http://reactivex.io/documentation/operators/concat.html) super explanatory.

NachoSoto replied:
I actually like your example too @neilpa!

kastiglione replied:
Yeah nice!

iv-mexx replied:
@neilpa I like that example :+1:

NachoSoto commented:
Looking really good!

kennethgunn commented:
I'm a RAC3 noob, and that is some awesome documentation.

iv-mexx commented:
@neilpa (writing this here because the other comment will be collapsed):

I agree that the [operator diagrams](http://reactivex.io/documentation/operators/concat.html) are really helpful. Do you feel like we should linke to those for each operator in this document? The Problem is the slightly different terminology there (Observable vs Signal)... Maybe just custom draw them eventually :speak_no_evil: 

I don`t know about other implementations, but RxJava for example has those diagrams directly in the Javadoc which is super helpful. But after a quick google search I'm afraid Xcode Quickhelp does not support images anyway...

neilpa commented:
> Do you feel like we should linke to those for each operator in this document?

No, I just like to point to them though cause

> Maybe just custom draw them eventually 

Probably something like this would need to happen if we want diagrams. I may look into this sometime in the near future (if my other half-baked RAC visualization idea doesn't pan out).

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,392 @@ code-style will be used.
+Side effects can be injected on a `SignalProducer` with the `on` operator without actually subscribing to it. 
+
+```Swift
+producer
````
Given the confusion in #2060 we should either capture the returned producer here or `start` it below.

iv-mexx replied:
I've updated the example and added a note about starting the resulting producer.

iv-mexx commented:
> But after a quick google search I'm afraid Xcode Quickhelp does not support images anyway...

Well, I guess thats going to change:

> Comments can include Markdown syntax to add rich text and embedded images that display in Xcode’s Quick Help.
[Introducing Swift 2][https://developer.apple.com/swift/]

iv-mexx commented:
Hey @neilpa, I've seen your tweet about http://neilpa.me/rac-marbles/ (which looks really great btw! :+1: ). 
There`s quite a long TODO list, do you want to keep this PR here open until the visualisation is done and link each operator to its visualisation? Or should I add links to only those operators that are already done?

neilpa commented:
I'd only add links to the completed ones for now, but not block the PR on finishing the rest. I'd still like to give this PR another pass myself, which I should hopefully get to later tonight.

> There`s quite a long TODO list

Yea, I'd also like to organize it a bit differently as well but kept it close to rxmarbles for now since it was only a half day project so far.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
 
 ### Observation
+
+Signals can be observed with the `observe` function. It takes a `Sink` as argument to which any future events are sent. 
```
'takes a Sink ~~as~~ argument'

neilpa replied:
'`Signal`s can be observed'

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+})
+```
+
+Alternatively, callbacks for the `Next`, `Error`, `Completed` and `Interrupted` events can be provided which will be called when a corresponding event occurs.
````
Can you append another sentence that each param is optional and only those events you care about need callbacks.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
 ### Injecting effects
 
+Side effects can be injected on a `SignalProducer` with the `on` operator without actually subscribing to it. 
```
Same here, that all the params are optional.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+### Lifting
+
+Signal operators can be _lifted_ to operate upon `SignalProducer`s instead using the `lift` operator.
```
~~instead using~~ with

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+Signal operators can be _lifted_ to operate upon `SignalProducer`s instead using the `lift` operator.
+In other words, this will create a new `SignalProducer` which will apply the given signal operator to _every_ signal created from the incoming `SignalProducer`s
+just if the operator had been applied to each signal yielded from `start()`.
```
Did you mean to hard break here?

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+let (signal, sink) = Signal<String, NoError>.pipe()
+signal
+    |> map { string in string.uppercaseString }
+    |> observe(next: { println($0) })
```
`|> observe(next: println)

iv-mexx replied:
Nice, I did not know about this a few weeks back :+1:

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+sendNext(sink, 4)     // prints 4
+```
+
 ### Reducing
````
Lets call this `Aggregating` and then add a second paragraph about `collect` here.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
 
 ## Combining signals
 
+These operators combine multiple signals into a single new signal.
```
'These operators combine values from multiple signals into a unified new signal.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
 ### Zipping
 
+The `zip` function combines values of two signals into pairs. The elements of any Nth pair are the Nth elements of the two input signals. That means the output signal will always wait for both input signals to send and output.
```
two (or more) signals

neilpa replied:
elements of the ~~two~~ input signals

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+sendNext(numbersSink, 0)    // nothing printed
+sendNext(numbersSink, 1)    // nothing printed
+sendNext(lettersSink, "A")  // prints (1, A)
```
prints (0, A)

neilpa replied:
I'm guessing copy/paste bug.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+sendNext(numbersSink, 0)    // nothing printed
+sendNext(numbersSink, 1)    // nothing printed
+sendNext(lettersSink, "A")  // prints (1, A)
+sendNext(numbersSink, 2)    // prints (2, A)
```
prints nothing

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+zip(numbersSignal, lettersSignal)
+    |> observe(next: { println($0) },
+          completed: { println("Completed") })
```
can you single line this - `|> observe(next: println, completed: { println("Completed") }`

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+combineLatest(numbersSignal, lettersSignal)
+    |> observe(next: { println($0) },
+          completed: { println("Completed") })
```
can you single line this - `|> observe(next: println, completed: { println("Completed") })`

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+sendNext(lettersSink, "A")  // prints (1, A)
+sendNext(numbersSink, 2)    // prints (2, A)
+sendCompleted(numbersSink)  // nothing printed
+sendNext(lettersSink, "B")  // prints (2, B)
```
prints (1, B)

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
 ## Flattening producers
 
-### Concatenating
+The `flatten` operators transforms a `SignalProducer`-of-`SignalProducer`s into a single `SignalProducer`.
```
The `flatten` operator~~s~~

neilpa replied:
I would reword this into a single sentence.

'The `flatten` operator transforms `SignalProducer`-of-`SignalProducer`s into a `SignalProducer` whose values are forwarded from the inner producer in accordance with the provided `FlattenStrategy`.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+The `flatten` operators transforms a `SignalProducer`-of-`SignalProducer`s into a single `SignalProducer`.
+There are multiple different semantics of the operation which can be chosen as a `FlattenStrategy`.
+
+To understand, why there are different strategies and how they compare to each other, take a look at this example and
```
another spurious line break

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
 ### Merging
+
+The `.Merge` strategy works by immediately forwarding every events of the inner `SignalProducer`s to the outer `SignalProducer`.
```
forwarding every ~~events~~ value

neilpa replied:
Lets add a sentence that any error sent on the outer producer or any inner producer is immediately sent on the flattened producer and terminates it.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+let producer = signalProducer
+    |> on(started: {
+        println("Started")
+        }, event: { event in
```
The indenting here is off

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+The `.Merge` strategy works by immediately forwarding every events of the inner `SignalProducer`s to the outer `SignalProducer`.
+
+```Swift
+let (numbersSignal, numbersSink) = SignalProducer<AnyObject, NoError>.buffer(5)
````
Not a fan of any object here. Can you pick either numbers or strings.

iv-mexx replied:
sure

iv-mexx replied:
Oh... of course it possible to send strings "1", "2', "3"... d'oh :dash:

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+signal
+    |> flatten(FlattenStrategy.Merge)
+    |> start(next: { println($0) })
```
`start(next: println)`

neilpa replied:
signal |> flatten(.Concat) |> start(next: println)

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+This implies, that values of a `SignalProducer` are only sent, when the preceding `SignalProducer` has completed
+
+```Swift
+let (numbersSignal, numbersSink) = SignalProducer<AnyObject, NoError>.buffer(5)
````
Same here, lets stick to just numbers or letters.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+### Concatenating
+
+The `.Concat` strategy works by concatenating the `SignalProducer`s such that their values are sent in the order of the `SignalProducer`s themselves. 
+This implies, that values of a `SignalProducer` are only sent, when the preceding `SignalProducer` has completed
```
I'd reword these sentences a bit:

The `.Concat` strategy is used to serialize work of the inner `SignalProducer`s. The first producer sent by the outer producer is started immediately. Each subsequent producer is not started until the preceding one has completed. Errors are immediately forwarded to the flattened producer.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+sendCompleted(sink)
+
+sendNext(numbersSink, 1)    // prints 1
+sendNext(lettersSink, "A")  // nothing printed
```
Flipping these would make for a slightly better example of the semantics.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+sendNext(lettersSink, "B")  // nothing printed
+sendNext(numbersSink, 3)    // prints 3
+sendNext(lettersSink, "C")  // nothing printed
+sendCompleted(numbersSink)  // prints A, B, C
```
Can you move this before sending "C" to better show the semantics.

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+signal
+    |> flatten(FlattenStrategy.Latest)
+    |> start(next: { println($0) })
```
`|> start(next: println)`

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+let (signal, sink) = SignalProducer<SignalProducer<AnyObject, NoError>, NoError>.buffer(5)
+
+signal
+    |> flatten(FlattenStrategy.Latest)
```
`flatten(.Latest)`

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+signal
+    |> flatten(FlattenStrategy.Concat)
+    |> start(next: { println($0) })
```
lets just one line this `signal |> flatten(.Concat) |> start(next: println)`

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+The `.Latest` strategy works by forwarding only events from the latest input `SignalProducer`.
+
+```Swift
+let (numbersSignal, numbersSink) = SignalProducer<AnyObject, NoError>.buffer(5)
````
Can we add a third signal here just to make the example even more clear

neilpa replied:
And lets stick to a single type here as well.

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+The `catch` operator catches any error that may occur on the input `SignalProducer`, then starts a new `SignalProducer` in its place.
+
+```Swift
````
I'd structure this example a bit differently to better showcase that catch is good at creating sensible defaults in the face of errors.

neilpa replied:
How about

```swift
let (producer, sink) = SignalProducer<String, NSError>.buffer()
let error = NSError(domain: "domain", code: 0, userInfo: nil)

producer
    |> catch { error in SignalProducer(value: "Default") }
    |> start(next: println)

sendNext(producer, "Once")
sendNext(producer, "Twice")
sendError(producer, error) 
```

iv-mexx replied:
Nice :+1:

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+
 ### Retrying
 
+The `retry` operator ignores up to `count` errors and tries to restart the `SignalProducer`.
```
The `retry` operator will restart the original `SignalProducer` on error up to `count` times.

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+
+```Swift
+let (signalA, sinkA) = SignalProducer<String, NSError>.buffer(5)
+let (signalB, sinkB) = SignalProducer<String, NSError>.buffer(5)
````
s/signal/producer/

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,394 @@ code-style will be used.
+var tries = 0
+let limit = 2
+let error = NSError(domain: "domain", code: 0, userInfo: nil)
+let signal = SignalProducer<String, NSError> { (sink, _) in
```
s/signal/producer/

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,394 @@ code-style will be used.
+```Swift
+let (numbersSignal, numbersSink) = SignalProducer<AnyObject, NoError>.buffer(5)
+let (lettersSignal, lettersSink) = SignalProducer<AnyObject, NoError>.buffer(5)
+let (signal, sink) = SignalProducer<SignalProducer<AnyObject, NoError>, NoError>.buffer(5)
````
s/signal/producer/

neilpa commented:
Looking good overall. A number of small pieces of feedback but nothing major.

iv-mexx commented:
@neilpa thanks for your review! I think I have incorporated all your comments (hope I didn't miss anything) and also added [RAC Marbles](http://neilpa.me/rac-marbles) links for operators that are already done there.

thbaja commented:
@iv-mexx: This is great - exactly what I have been looking for! I'm totally new to RAC and this helps a lot.  Thanks!

Currently I've been combining info from the changelog, framework overview, test cases and some more or less outdated tutorials and github projects to navigate through the new syntax and paradigmes.

I think it would be nice to include this in the release notes as well, since that is more or less the source people use to learn more.

Here is a direct link for other people who are lucky enough to stumble across this thread:
https://github.com/iv-mexx/ReactiveCocoa/blob/feature/operators/Documentation/BasicOperators.md

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,416 @@ code-style will be used.
+```
+
+
+The `reduce` operator is used to aggregate a signals values into a single combine value. Note, that the final value is only sent after the source signal completes.
````
Lets keep reduce before collect since it's the more general version.

iv-mexx replied:
Makes sense :+1:

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,416 @@ code-style will be used.
+Alternatively, callbacks for the `Next`, `Error`, `Completed` and `Interrupted` events can be provided which will be called when a corresponding event occurs.
+
+```Swift
+signal.observe(next: { next in
````
In the RAC code we follow the convention of not indenting the callback params

```swift
signal.observe(next: { next in
    println("Next: \(next)")
}, error: { error in
    println("Error: \(error)")
}, completed: {
    println("Completed")
}, interrupted: {
    println("Interrupted")
})

neilpa commented on Documentation/BasicOperators.md:
```diff
@@ -46,41 +46,416 @@ code-style will be used.
+            println("Disposed")
+        }, next: { next in
+            println("Next: \(next)")
+    })
```
Can you un-indent these ones too

iv-mexx replied:
Is it correct at the same indentation level as the `|>` or should it be completely to the left?

neilpa replied:
At the same level of `|>`

neilpa commented on Documentation/BasicOperators.md:
````diff
@@ -46,41 +46,416 @@ code-style will be used.
+The `mapError` operator transforms errors in the signal to new errors. 
+
+```Swift
+struct CustomError: ErrorType {
````
The more common practice will be going from `NSError` to `CustomError` so lets invert the mapping in this example

```swift
enum CustomError: ErrorType {
   case Foo = "Foo"
   case Bar = "Bar"
   case Other = "Other"

   var nsError: NSError {
       return NSError(domain: "CustomError.\(rawValue)", code: 0, userInfo: nil)
   }

   var description: String {
       return "\(rawValue) Error"
   }
}

let (signal, sink) = Signal<String, CustomError>.pipe()

signal
    |> mapError { error in
        switch error.domain {
        case "com.example.foo": return .Foo
        case "com.example.bar": return .Bar
        default: return .Other
        }
    }
    |> observe(error: println)

sendError(sink, NSError(domain: "com.example.foo", code: 42, userInfo: nil) // prints "Foo Error"
```

iv-mexx replied:
:+1:  I had to add a few explicit type annotations to make the example run in the playground.

neilpa commented:
Just a couple more comments and I'm ready to merge this.

iv-mexx commented:
@neilpa Thanks again, just the one quick question about the indentation for the _Injecting effects_ example.

neilpa commented:
Indent at the same level as `|>`, in case that comment got buried

iv-mexx commented:
Ok, thats what I did, so it should be good to go :star2:

neilpa commented:
:ship: :tada:
